pandas/core/reshape/melt.py

```diff
@@ -133,8 +149,9 @@ def melt(
             mdata[col] = np.tile(id_data._values, K)
 
     mcolumns = id_vars + var_name + [value_name]
-
-    if frame.shape[1] > 0:
+    if frame.shape[1] > 0 and not any(
+        isinstance(dt, pd_dtypes) for dt in frame.dtypes.values
```

Comment:
You can actually use `lib.is_np_dtype`. The current check would miss any future new dtype

Reply:
Cool! This is good to know for future PRs

Reply:
Upon researching and trying this method out further it's still returning False when I try to use it and therefore going to the wrong side of the conditional to fix the bug.  I think because of this in the function body of `lib.is_np_dtype`.  The dtypes in the test case are `<class 'pandas.core.dtypes.dtypes.DatetimeTZDtype'>`:
```Python
def is_np_dtype(object dtype, str kinds=None) -> bool:
    if not cnp.PyArray_DescrCheck(dtype):
        # i.e. not isinstance(dtype, np.dtype)
        return False
    if kinds is None:
        return True
    return dtype.kind in kinds
```

Reply:
Well I think this can still be used to make the condition take the right path since the pseudo-logic  should be

```
if there are any pandas dtypes:
    take the concat path
else:
    take the raval path
```

?

Reply:
Gotcha!  I just needed to invert the logic.  I really appreciate you clarifying this for me! 😎

Reply:
So after working with this more I apparently needed to check for `is_integer_dtype(dt)`.  Simply checking for numpy `dtypes` was failing for `test_melt_ea_columns` and `test_melt_ea_dtype`.

Reply:
I am suspicious that `is_integer_dtype` is an appropriate check here; this code manipulation doesn't seem to be integer specific.

Reply:
It has to do with how the current implementation of `concat` handles `DatetimeTZDtype`.  

If you call `concat.values` like in the current `melt` implementation you get:
```Python
array(['2023-02-28T15:00:00.000000000', '2023-02-28T15:00:00.000000000',
       '2023-03-09T15:00:00.000000000', '2023-03-10T15:00:00.000000000'],
      dtype='datetime64[ns]')
```
If the desired outcome is to have the format is to have it presented like: `"2023-03-01 00:00:00+0900"`, then I'm wondering if we should simply check for  `DatetimeTZDtype` like I did originally?

Reply:
Maybe the `.values` call is outdated and not needed anymore as merging with this type has gotten better over the years

Reply:
Removing the `.values` call would work for only `DataFrames` that have `DatetimeTZDtype`.  I tried removing it for everything but it failed 17 of the current test cases.  Should I add in a conditional to remove that call only for `DatetimeTZDtype`?

Reply:
Ideally there should be no dtype-specific checking here since this function just reshapes. Generally numpy data types have 2D support while pandas dtypes have only 1D support so possibly this branch needs to account for that

Reply:
But `DatetimeTZDtype` is a 2D structure which is why this bug is happening as noted in `core/internals/block.py`.  I'm not sure how to proceed but am open to more suggestions.
```Python
    def _validate_ndim(self) -> bool:
        """
        We validate dimension for blocks that can hold 2D values, which for now
        means numpy dtypes or DatetimeTZDtype.
        """
        dtype = self.dtype
        return not isinstance(dtype, ExtensionDtype) or isinstance(
            dtype, DatetimeTZDtype
        )
```

Reply:
Ah I see. I think pandas dtypes support a `_supports_2d` attribute that you could check here

Reply:
That did it! I combined this with`is_np_dtype` your previous answer in order to get this to work(since np dtypes don't have the `_supports_2d` attribute. 

I really appreciate your Time here Matthew, I learned a lot more about the private APIs in this PR.

Reply:
@mroeschke I changed the conditional to:
```Python
if frame.shape[1] > 0 and not any(
        hasattr(dt, "_supports_2d") and dt._supports_2d for dt in frame.dtypes.values
    ):```

I figured it would be a little clearer this way and it also eliminates any `dtype` checks.